Add improvements to heuristic for bike rental and car modes (Fixes #2671)

evansiroky opened the pull request:
To be completed by pull request submitter:

- [x] **issue**: #2671.

This PR seeks to be the first of a few PRs that are being split out of #2664.  This PR is just the improvements made to the InterleavedBidirectionalHeuristic that fixed a few bugs and improved the speed of car routing searches.

This heuristic is being used as a way to speed up searches with transit by attempting to traverse and score ways close to the origin and destination before doing the main graph search. A number of problems were identified in this issue here that were the result of problems with the InterleavedBidirectionalHeuristic. There was an issue with the heuristic not adding bike rental stations or car rental stations, thus ignoring them in searches. One-way streets were being traversed backwards while walking a bike which led to trips that did not respect the maxWalkDistance parameter. Also, trips with cars were traversing the entire graph, so now, if the car mode is enabled, the pretransit search isn't done at all and a limiter on the posttransit search was implemented to stop the search once the origin was reached. And finally, forked states weren't being analyzed so that was fixed.

- [x] **roadmap**: Check the [roadmap](https://github.com/orgs/opentripplanner/projects/1) for this feature or bug. If it is not already on the roadmap, PLC will discuss as part of the review process.
- [x] **tests**: Have you added relevant test coverage? Are all the tests passing on [the continuous integration service (Travis CI)](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Developers-Guide.md#continuous-integration)?
- [x] **formatting**: Have you followed the [suggested code style](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Developers-Guide.md#code-style)? 

To be completed by @opentripplanner/plc:

- [ ] reviews and approvals by 2 members, ideally from different organizations
- [x] **before merging**: add a bullet point to the [changelog file](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Changelog.md) with description and link to the linked issue
- [ ] **after merging**: update the relevant card on the [roadmap](https://github.com/orgs/opentripplanner/projects/1)

evansiroky commented:
@optionsome I just added a commit that I think will resolve the issue you are seeing.  Let me know if it fixed your issue.

optionsome commented:
Sorry seems like the latest commit did not change anything in our context. However, I can see that thanks to the earlier commit, the bike rental is actually being used more.

I should elaborate our issue a bit more. Users are often more willing to cycle more than they want to walk. Therefore, in our UI, we have defined different "maxWalkDistance" values when BICYCLE mode is enabled or disabled. However, with BICYCLE_RENT it gets a bit more complicated as on average the amount of walking is more than with BICYCLE mode. In some cases, there are no bicycle rental stations anywhere near by so the trip does not use any cycling. So I think there are three possible solutions -- we could either ignore the maxWalkDistance limit in the cycling done on a rental bicycle, we could have separate "maxCyclingDistance" value that limits the amount of cycling to a transit stop or my least favourite option, we could increase the maxWalkDistance with some multiplier if BICYCLE_RENT mode is enabled. There are potentially other solutions that I haven't thought of.

evansiroky commented:
@optionsome Is there a URL of a trip planning website that I can play around with and see these results?

If no bike park or bicycle rental stations are within the max walk distance of the origin, then OTP seems to be working correctly by being unable to find a viable route that uses a bicycle rental or bicycle parking.  I do like the option of having separate parameters for max walking and max cycling.

However, since these changes already seem to be improving your results is that alone enough to earn your approval for this PR?

optionsome commented:
In our case, it is not enough that the bicycle rental station, where you pick up the bicycle, is within the maxWalkDistance. It seems like the transit stop from where your trip continues needs to be in that range. Unfortunately our rental bicycle stations are turned off for the winter season so there is no OTP instance hosted online where you could test this. I uploaded the graph and the router config necessary to test this here: https://gtfsdata.blob.core.windows.net/graph/hsl.zip. Also, I pushed the branch I used to build the graph and to test these changes here: https://github.com/HSLdevcom/OpenTripPlanner/tree/bicycle_rent_test. That branch is based on upstream's master and it has your commits + couple of commits to hardcode bicycle availability. For example, by changing the maxWalkDistance value in this query
``` http://localhost:8888/otp/routers/hsl/plan?maxWalkDistance=2000&fromPlace=60.23572%2C24.81391&mode=BICYCLE_RENT%2CRAIL&numItineraries=1&time=14%3A00&date=2018-11-26&toPlace=60.16790%2C24.93639 ```
you can notice how the maxWalkDistance needs to be in the range on the train station instead of the nearest bicycle rental station. I attached an image where the origin is inside the red circle and the rental station inside the blue circle. They are basically next to each other but the train station is almost 2000 meters away from the origin.
![rental](https://user-images.githubusercontent.com/24825005/48936003-2ee4da80-ef12-11e8-9807-9951957b3742.PNG)
You can see a map of the rental stations in https://reittiopas.hsl.fi/ if you zoom close enough but you first need to enable that map layer from the bottom left layer menu.

I don't think solving these issues for us was in the original scope of this pull request and the changes you made only seemed to have a positive effect, so I can give you an approval on those changes alone. However, I don't mind if you quickly find a solution that fixes this aforementioned issue :).

kalon33 commented:
@optionsome do you need a particular OTP version to perform the tests you can't perform using HSL data because bike rental stations are closed? I've a OTP instance on Paris running latest hsldevcom/OpenTripPlanner master and including active bike rental stations.

optionsome commented:
I don't think the OTP version matters too much to prove that the issue exists because it has been existing for a while now. However, the type of bike rental station API might affect the results as we are using a dynamic updater and not sure if that is available for your use?

kalon33 commented:
@optionsome I'm currently using the GBFS dynamic updater for bike sharing stations, is that relevant for you?

optionsome commented:
You can try to see if you can replicate the issue with that setup. The steps to replicate:
1. Pick BICYCLE_RENT mode and some transit mode where it is possible to carry the bicycle with you
2. Pick origin that is further away from any viable transit stop than maxWalkDistance but there is a bicycle rental station within the maxWalkDistance range

kalon33 commented:
@optionsome I don't know if I properly understood what you are trying to observe exactly. Maybe that would be quicker for you to replicate it by yourself, than me trying that tomorrow: https://paris.acolytesanonymes.org . Hope that it will still help.

optionsome commented:
Thank you for that link. Did some quick testing and does not seem like the same issue is affecting you. Although, seems like the maxWalkDistance affects how much BICYCLE_RENTAL mode is being used. For example, https://paris.acolytesanonymes.org/otp/routers/default/plan?maxWalkDistance=2000&fromPlace=48.84845%2C2.32635&mode=BICYCLE_RENT%2CRAIL&numItineraries=1&time=14%3A00&date=2018-11-26&toPlace=48.87330%2C2.28996
vs
https://paris.acolytesanonymes.org/otp/routers/default/plan?maxWalkDistance=1000&fromPlace=48.84845%2C2.32635&mode=BICYCLE_RENT%2CRAIL&numItineraries=1&time=14%3A00&date=2018-11-26&toPlace=48.87330%2C2.28996

evansiroky commented:
@optionsome Yeah, the desired transit station must also be within the range of maxWalkDistance.  This is because traveling while walking or bicycling both increment the walking distance:

https://github.com/opentripplanner/OpenTripPlanner/blob/0cfb77d4689c0b1fa659ddcf37de889f1c44825f/src/main/java/org/opentripplanner/routing/edgetype/StreetEdge.java#L469-L471

As a result, whenever the remaining weight is being estimated, anything falling outside of the maxWalkDistance will automatically be discarded from the graph search:

https://github.com/opentripplanner/OpenTripPlanner/blob/0cfb77d4689c0b1fa659ddcf37de889f1c44825f/src/main/java/org/opentripplanner/routing/algorithm/AStar.java#L179-L185

We observed this in our OTP instance and decided instead of adding too many parameters, we'd just have a default value of 3 miles (4.8km) in the UI when the bicycling + transit modes were being used together.

As I mentioned before I think it would make sense to have separate parameters for the max cycling distance and walking distance and perhaps reset the counter when transit is boarded, but that seems like a bigger change beyond the scope of this PR.

optionsome commented:
Thanks for the clarification that this is indeed the intended behavior at the moment and I agree that changing this behavior is not in the scope of this PR. We should either continue the conversation in  https://github.com/opentripplanner/OpenTripPlanner/issues/2311 or close that, as it is old and originally addresses a bit different issue, and create a new issue.

optionsome commented:
Did some testing with more test cases while using BICYCLE_RENT and CAR_PARK modes together with WALK and TRANSIT. Seems like the BICYCLE_RENT performance was about the same as before, maybe slightly worse. However, I could see that the number of itineraries returned (searched for 5 itineraries) with CAR_PARK mode went down by quite a lot after these changes. In 400 requests,  the version with these changes returned less itineraries in 168/400 requests than without changes which returned less itineraries in 41/400 requests. Requests were slightly faster with these changes. I have to admit that my testing setup's performance changes a bit randomly but it shouldn't cause this large effects. Are these changes expected or an unknown side effect? I can try to redo the tests some other day if you are not able to reproduce these results.

evansiroky commented:
@optionsome How do I setup your testing environment?  Is there a testing script I can use?

optionsome commented:
Sorry for the delay. I redid my tests and saw positive effects on any other BICYCLE or CAR related mode, but for some odd reason, less itineraries were returned after these commits when CAR_PARK mode was selected. However, routing was still faster for even that mode.

The CAR_PARK implementation I used does not exist in upstream unfortunately. I used  https://github.com/HSLdevcom/OpenTripPlanner/tree/test-master branch to test without these changes and https://github.com/HSLdevcom/OpenTripPlanner/tree/bike-rental-fix-test to test with these changes. The OSM data I used to build the graph can be found https://api.digitransit.fi/routing-data/v2/hsl/hsl.pbf and the GTFS data is https://api.digitransit.fi/routing-data/v2/hsl/HSL.zip.

router-config.json
```
{
  "updaters": [
    {
      "type": "bike-rental",
      "sourceType": "smoove",
      "url": "http://helsinki-fi.smoove.pro/api-public/stations",
      "frequencySec": 10
    },
    {
      "type": "car-park",
      "sourceType": "hsl-parkandride",
      "url": "http://p.hsl.fi/api/v1/facilities.json?limit=-1",
      "frequencySec": 600
    }
  ]
}
```

build-config.json
```
{
  "areaVisibility": true,
  "staticParkAndRide": false,
  "parentStopLinking": true
}
```

I used HSL [fork](https://github.com/hsldevcom/OTPQA) of OTPQA for testing. First, to do a test run that generate output files that can be used for comparisons, you need to run `python otpprofiler.py -c 400 -i 5 -o -m 'CAR_PARK,TRANSIT' http://localhost:8888/otp/routers/hsl/`.
It tries to search for 5 itineraries using 'CAR_PARK,TRANSIT' modes and http://localhost:8888/otp/routers/hsl/ is the endpoint where I had my OTP running. You get a run_summary file that you can rename to anything you want. If you run the otpprofiler.py again against some different version of OTP, you can compare those run_summary files by running `python compare.py -i <first_file>.json <second_file>.json`.
-i lets you see the change in number of itineraries returned in each request and in total. You can also use -p to get an idea how fast the queries executed. Each query that differs somehow between those two run_summaries will be printed and you can see by how much they differ (in the number of itineraries, for example) and also an ID for the query you can use to find the specific query in those run_summary files.

Unfortunately, I don't know how you can generate the required request data (requests.json) for any other area. The upstream development of OTPQA has been stagnant in the recent years, so I'm not sure what other people are using for testing the performance of OTP.

evansiroky commented:
Thanks for the info, @optionsome.  I'll investigate tomorrow.

evansiroky commented:
@optionsome I was able to run OTPQA as you mentioned and did receive output.  I also noticed a lot of searches returning less itineraries than before.  After looking more into the behavior of OTP, I realized that my code changes were based on a flawed assumption that there are no limitations on how much of a graph can be explored in this heuristic initialization when the car mode is enabled.  Therefore, I have reverted a bunch of my changes and added in some comments to help explain what is happening.

Also, regarding the Park and Rides, I added in some code that will treat a ParkAndRideVertex as a pre-transit vertex in this heuristic.  I believe this is the correct interpretation of the intent of this algorithm because any P&R that is within the max pretransit time should be prioritized more than those falling outside of the max pretransit time.  I looked in detail at the resulting before-and-after differences in a few results and it seemed like sometimes a P&R closer to the origin was used while other times one closer to the destination was used, but generally the travel times seemed faster overall.

evansiroky commented:
A drop in the number of itineraries should be expected as a result of this pull request for a few reasons.

With bicycling, the changes here: https://github.com/opentripplanner/OpenTripPlanner/blob/08effdf265b958083b97c2d12a113fb6f8ebb194/src/main/java/org/opentripplanner/routing/algorithm/strategies/InterleavedBidirectionalHeuristic.java#L112-L116
actually fixes a bug in how this heuristic is initialized which was causing bicycling and walking trips to exceed the max walk distance in areas that have one-way streets.  Therefore, I would expect that more trips would now not be possible with the default limit of 2km of bicycling/walking before transit.

Shown below is a picture of the pretransit vertices in a TRANSIT,BICYCLE trip generated before these changes using a Portland graph.  The diamond-shaped area on the top is the expected vertices within the maxWalkDistance.  However, extra unexpected vertices on one-way streets were also found that extended the allowable bicycling distance of the trip well beyond the maxWalkDistance.

<img width="625" alt="screen_shot_2018-10-05_at_10 58 07_am" src="https://user-images.githubusercontent.com/3112493/50267689-eddec980-03dc-11e9-9f66-a1ecc20744a4.png">

After these changes, the resulting vertices retained only the diamond shape as should be expected.

The interpretation of P&Rs is also being corrected (along with bike park and bike rental) by adding lines 156-158:  

https://github.com/opentripplanner/OpenTripPlanner/blob/08effdf265b958083b97c2d12a113fb6f8ebb194/src/main/java/org/opentripplanner/routing/algorithm/strategies/InterleavedBidirectionalHeuristic.java#L154-L183

The only way that P&Rs were even considered in searches in the first place is due to the fact that they are not a StreetVertex.  Therefore, when their estimated remaning weight is evaluated they always got assigned either the `maxWeightSeen` or a `postBoardingWeight` in the search if it happened to be close to the target vertex.  Previously what was probably happening was that since any P&R was assigned the `maxWeightSeen` they were getting inserted so far down in the priority queue that the very closest P&R would be evaluated first and then the ensuing transit search from that P&R would be prioritized before almost all other P&R evaluations could occur.

Consider the folloiwng use case:  A suboptimal P&R exists close to the origin, but a more optimal P&R exists slightly further away.  Before these changes, the closest P&R was heavily prioritized and chosen first.  The total travel time of this itinerary is 58 minutes.

<img width="536" alt="screen shot 2018-12-19 at 10 34 11 pm" src="https://user-images.githubusercontent.com/3112493/50268107-58443980-03de-11e9-8601-44f35981fbcf.png">

After the changes, all P&Rs have an estimated remaining weight of 0 and are therefore inserted in the priority queue at about the same position as the rest of the StreetVertices that are nearby.  The more optimal P&R further away now rises to the top of the results with its superior travel time of 28 minutes.

<img width="495" alt="screen shot 2018-12-19 at 10 38 37 pm" src="https://user-images.githubusercontent.com/3112493/50268238-df91ad00-03de-11e9-9f13-ca2bbe674bff.png">

The reason why this may result in fewer itineraries is that more park and rides are being reached which causes more graph searching into the transit network which can result in slightly slower graph search times, but I believe an overall improvement in the quality of the results.

optionsome commented:
You made valid points. Did you generate that first image with QGIS? How did you import the vertices? Would probably have to take a better look at what the itineraries, that are missing after these changes, look like to make sure that the reason for these changes are those that you pointed out. Hopefully I will have time for that at the start of next year.

evansiroky commented:
I had to manually debug OTP and insert the following code to write out the vertices to a csv.

```java
CsvWriter pretransitVertexWriter = new CsvWriter("pretransit-vertices.csv");
try {
    pretransitVertexWriter.writeRecord(new String[]{
        "name",
        "type",
        "latitude",
        "longitude"
    });
    for (Vertex preTransitVertex : preTransitVertices) {
        pretransitVertexWriter.writeRecord(new String[]{
            preTransitVertex.getName(),
            preTransitVertex.getClass().getName(),
            String.valueOf(preTransitVertex.getLat()),
            String.valueOf(preTransitVertex.getLon())
        });
    }
} catch (IOException e) {
    LOG.info("failed to write something to pretransit vertices");
}
```

Sometimes I had to add a breakpoint because the heuristic was being called again and overwriting stuff.

abyrd commented:
Hello @optionsome and @evansiroky, I just read through this discussion, and one thing I'd like to clarify is that in current OTP, there is no actual walk/bike distance limiting feature. The walk distance parameter does one specific thing: it masks off an area of the street graph that is useable. Everything within that radius of the origin and destination is considered in routing; everything else is not.

With the optimization algorithm we are using, it is only acceptable to truncate paths based on variables that are being minimized. We are minimizing generalized cost alone. Therefore you can only truncate paths based on generalized cost. A max walk or bike limit truncates paths based on walking or biking distance, which violates basic assumptions of the algorithm.

You can only cut off paths based on walk distance if you modify the search algorithm to find all Pareto-optimal paths representing all trade-offs between generalized cost and walk distance. Otherwise behavior is undefined, and has been seen to produce very strange and difficult to debug results in the past.

Because OTP optimizes generalized cost, if you want less walking or biking, the preferred (and only reasonably correct) approach is to increase the cost of walking or biking (the "reluctance" parameters).

evansiroky commented:
I added another commit that will speed up park and ride searches.  Prior to this commit, main searches would end up exploring a significant amount of vertices in walk mode since there is no limit to the max walk distance in the main graph search.  All pretransit vertices were fair game and a significant majority of them included vertices only reachable in car mode up to the maxPretransitTime. 
 The change here preserves the maximum walking distance in the main graph search and serves as a safegaurd against undesireably long walking legs.

optionsome commented on src/main/java/org/opentripplanner/routing/algorithm/strategies/InterleavedBidirectionalHeuristic.java:
```diff
@@ -120,9 +128,11 @@ public void initialize(RoutingRequest request, long abortTime) {
+        // for trips near the cutoff.  However, if the car mode is enabled, excessive walking legs
+        // can occur that take advantage of the vertices explored only by car, so keep the cap if
+        // the car mode is enabled.
+        if (!request.modes.getCar()) request.setMaxWalkDistance(Double.POSITIVE_INFINITY);
```
I think that these one liner if statements should be avoided according to the code style convention. Side note, the link for the sun code style convention is outdated [here](http://docs.opentripplanner.org/en/latest/Developers-Guide/#java). I think the new link is perhaps https://www.oracle.com/technetwork/java/codeconventions-150003.pdf.

optionsome replied:
Not sure if that link should be updated in this PR or in a separate pull request

evansiroky commented:
I have yet again extensively refactored this heuristic.  @abyrd and I had a few work sessions together and he gave some good review along the way (see https://github.com/opentripplanner/OpenTripPlanner/commit/d1f68aecf20c658a9b374e3471b8f736a9a009a4#comments).  

The largest change is that this heuristic now returns an estimated remaining weight based off of the current non-transit mode.  It also is now hopefully future-proofed against issues that were observed by not checking enough vertex types the right way (ie with BikeParkVertex and ParkAndRideVertex types).  Many more comments were added that hopefully clarify what the heuristic does.  Also, the code in the estimatedRemainingWeight method has been reorganized and a few variables renamed to hopefully have more clarity in what the purpose of each variable is.

abyrd commented:
Hi @evansiroky, can you confirm that this is still the first PR in the series for car rental / TNC routing, and that it's up to date with any new work on your local branches? Now that we have merged #2714 we can start merging this and the CS Flex routing into the `dev-1.x` branch.

evansiroky commented:
Yep, this is still the first PR, there'll be more to come. I just updated to the latest in dev-1.x.

optionsome commented on src/main/java/org/opentripplanner/routing/algorithm/strategies/InterleavedBidirectionalHeuristic.java:
```diff
@@ -132,42 +223,65 @@ public void initialize(RoutingRequest request, long abortTime) {
         final Vertex v = s.getVertex();
-        if (v instanceof StreetLocation) {
-            // Temporary vertices (StreetLocations) might not be found in the street searches.
+        if (v instanceof StreetLocation || v instanceof TemporaryVertex) {
```
I saw you had already helped @abyrd understand this part but could you still elaborate how this check replaces the instanceof checks you had here earlier. For example, BikeRentalStationVertex itself isnt a TemporaryVertex but there is always(?) TemporaryVertex linked to it or how does this work?

evansiroky replied:
Here's a diagram of the inheritance of the vertex types that now includes the vertex types introduced from the GTFS-FLEX work.

![otp vertex inheritance 1](https://user-images.githubusercontent.com/3112493/53056974-23964200-3462-11e9-93bb-29e2e0b2d507.png)

After discovering that the GTFS-FLEX vertices inherit from the TemporaryVertex I went ahead and changed the ordering of the vertex type checks in this PR. There are 5 possibilities that need to be checked for. The first is checking if a vertex is an instance of a TransitVertex in which case it should lookup the remaining weight fron the transit vertices. The second checks if a vertex is a StreetLocation or TemporaryVertex which need to be searchable but sometimes are not discovered during the streetSearch method. The third is checking whether the search has not yet boarded transit in which case if a preTransitVertex has been found, the remaining weight should be 0, but not Infinity. Fourthly, a check is needed to see if the search has boarded transit and is near the destination in which case a remaining weight estimate can be returned. And finally if none of those options have been found, the vertex must be a StreetVertex that is not close enough to the origin or destination to warrant further examination.

I hadn't seen Entur's fork and change. I think what this is doing is allowing the streetSearch method to explore nearby streets and thus other stops if the starting vertex happens to be a transit stop.

abyrd replied:
Thanks @evansiroky for adapting to the flex transit changes, and the inheritance diagram is very helpful in understanding the situation.

evansiroky commented:
I'm a little unsure why some requests would be slower if they're returning the same results. There are a few things that consume more resources such as the use of a HashMap for storing VertexModeWeight objects instead of a TObjectDoubleMap. How much slower are the requests and is there a certain pattern to them?

evansiroky commented:
Updated to the latest `dev-1.x` branch again. Could somone from the @opentripplanner/plc give a review? @barbeau, @clinct, @drewda, @fpurcell, @gmellemstrand, @irees, @novalis, @optionsome, @pailakka, @sdjacobs, @sheldonabrown, @t2gran?

optionsome commented:
I ran OTP with --visualize now. I used relatively large maxWalkDistance value. For these example images I used `min=QUICK&walkSpeed=1.222&arriveBy=True&maxWalkDistance=10000&fromPlace=60.248396%2C25.011233&mode=BICYCLE_RENT%2CTRANSIT&numItineraries=5&time=14%3A00&date=2019-02-25&toPlace=60.204534%2C24.889584&typical=False` as query parameters.
This is how it looks like without this pr
![master_visualized](https://user-images.githubusercontent.com/24825005/53408608-92761c80-39c7-11e9-83e8-f89bfceae470.PNG)
and this is how it looks like with these changes
![pr_visualized](https://user-images.githubusercontent.com/24825005/53408627-9f930b80-39c7-11e9-95a5-5c875d6b8ebd.PNG)
Routing is slower with these changes as more edges are explored. Without these changes it's able to find more itineraries faster and with less walking.

Forgot to mention in my last feedback but the comments you have added in the code are really helpful in making it easier to understand the code.

evansiroky commented:
Oh man, that is not good and not what I expected. I will investigate this further to see why so many edges are being explored.

evansiroky commented:
After spending numerous hours trying to reproduce exactly what you are seeing, I am unable to do so for the results you are seeing before this PR.

First off, I don't understand how you got the visualizer to accept a `BICYCLE_RENT` mode. Also, I am not sure what the `typical=False` parameter is supposed to mean.

In my own attempts to reproduce this, I tried using the method I mentioned in (https://github.com/opentripplanner/OpenTripPlanner/pull/2677#issuecomment-449066300). This is what I observed with this request: http://localhost:8888/otp/routers/default/plan?fromPlace=60.24844263880455%2C25.011502504348755&toPlace=60.20471876703638%2C24.88950490951538&time=2%3A30pm&date=02-27-2019&mode=TRANSIT%2CWALK%2CBICYCLE_RENT&maxWalkDistance=9656.039999999999&arriveBy=true&wheelchair=false&optimize=TRIANGLE&triangleTimeFactor=1&triangleSlopeFactor=0&triangleSafetyFactor=0&locale=en

As expected, I am seeing differences in the pre and post-transit vertices in the PRs due to the extra traversal of one-way streets.

Pretransit vertices before this PR:

<img width="862" alt="screen shot 2019-02-27 at 11 44 33 am" src="https://user-images.githubusercontent.com/3112493/53518169-4db0bb00-3a85-11e9-8e58-02c48efac2fd.png">

Pretransit vertices ater this PR:

<img width="862" alt="screen shot 2019-02-27 at 11 45 31 am" src="https://user-images.githubusercontent.com/3112493/53518166-4db0bb00-3a85-11e9-897a-6cb97a85f0cd.png">

Post-transit vertices before this PR:

<img width="862" alt="screen shot 2019-02-27 at 11 44 50 am" src="https://user-images.githubusercontent.com/3112493/53518168-4db0bb00-3a85-11e9-9988-80723ed88da3.png">

Post-transit vertices after this PR:

<img width="862" alt="screen shot 2019-02-27 at 11 45 43 am" src="https://user-images.githubusercontent.com/3112493/53518165-4db0bb00-3a85-11e9-9f99-f0e8b5d04a47.png">

I am also able to see the approximate outline of your second image with the transit vertices that were explored, however I am seeing this _both_ before and after this PR.

Post boarding weights after all searches before this PR:

<img width="863" alt="screen shot 2019-02-27 at 12 19 55 pm" src="https://user-images.githubusercontent.com/3112493/53522619-a422f700-3a8f-11e9-8899-fca835a0ab38.png">

Transit vertices after this PR:

<img width="862" alt="screen shot 2019-02-27 at 12 46 07 pm" src="https://user-images.githubusercontent.com/3112493/53522618-a422f700-3a8f-11e9-969b-fd44330b1e04.png">

As can be seen, pretty much all of the transit network is being explored both before and after this PR. However, I also notice that the HSLdevcom fork has the ability to modify the `HEURISTIC_STEPS_PER_MAIN_STEP` variable. Are you by chance using the HSLDevCom fork and setting this to a lower value than 8 in the tests of the results before this PR? Regardless, it sure seems like 8 is too large of a number since basically the entire transit network was explored for a trip that's only 11km by bike.

optionsome commented:
Sorry, I really should have given real steps/data to reproduce the issue. The problem can be reproduced with upstream code (just a slight modification needed to get Smoove bike rental updater to work in offseason). Didn't realise there was typical=False in the query but as far as I know it does nothing. It was just there because I picked some example query that behaved differently after running https://github.com/hsldevcom/OTPQA and I think OTPQA just adds it there so you can differentiate queries from "real ones".

I think you were trying to create the query from visualizer UI itself? You can do whatever query you want to the API and have the visualizer UI open and you will be able to see what OTP does. I don't personally use the visualizer often so I don't know all the things you can do with it nor how buggy it really is but I think in this case it might give a somewhat accurate visualization of what is happening.

router-config.json
```
{
  "updaters": [
    {
      "id": "hsl-bike-rental",
      "type": "bike-rental",
      "sourceType": "smoove",
      "url": "http://helsinki-fi.smoove.pro/api-public/stations",
      "frequencySec": 10
    }
  ]
}
```
build-config.json
```
{
  "areaVisibility": true,
  "staticParkAndRide": false,
  "parentStopLinking": true
}
```
Then you need to hardcode bicycle availability, you can use this code https://github.com/HSLdevcom/OpenTripPlanner/commit/e210edeb5a2a3564349cbb909658001208efe8fc.

I have uploaded the GTFS and OSM data I used [here](https://gtfsdata.blob.core.windows.net/graph/hsl.zip)

I built the graph with `java -jar -Xmx8G target/otp-1.4.0-SNAPSHOT-shaded.jar --build ./graphs/hsl` and ran OTP with `java -jar -Xmx8G target/otp-1.4.0-SNAPSHOT-shaded.jar --server --port 8888 --securePort 8889 --basePath ./ --graphs ./graphs --router hsl --visualize`

I got rid of some of the unnecessary parameters from the query I gave earlier. I was able to reproduce it with http://localhost:8888/otp/routers/hsl/plan?arriveBy=True&maxWalkDistance=10000&fromPlace=60.248396%2C25.011233&mode=BICYCLE_RENT%2CTRANSIT&numItineraries=5&time=14%3A00&date=2019-02-25&toPlace=60.204534%2C24.889584

Seems like arriveBy=True is one key factor in causing this problem in this case but I'm not sure there is some problem in handling arriveBy=True queries in general.

optionsome commented:
While working on some non-relevant feature today I noticed RoutingRequest reversedClone method has
```
ret.reverseOptimizing = !ret.reverseOptimizing; // this is not strictly correct
ret.useBikeRentalAvailabilityInformation = false;
```
Not sure if this is somehow relevant to this issue.

evansiroky commented:
Thanks for the additional bike rental code changes and how you're using this --visualize command. I'm now seeing what you are seeing and will do some more investigation. There are definitely more vertices being explored in the Astar search with this updated heuristic. I'll try to find out why.

evansiroky commented:
Just providing an update here, this may be ready for review, or I may continue to implement yet another idea I have for improving this heuristic.

The first thing I discovered is that the post-transit weights were being stored incorrectly, so I reimplemented that in https://github.com/opentripplanner/OpenTripPlanner/pull/2677/commits/61a28c0cad658203213761972846a9099e54fd0b.

However, even after doing that, the number of vertices being explored was still more than in the dev-1.x branch for request with `BICYCLE_RENT,TRANSIT` in https://github.com/opentripplanner/OpenTripPlanner/pull/2677#issuecomment-468225356. As I wrote previously, this was expected, but after examining exactly how many more vertcies would be explored, I found it to be about 7x the amount of vertices being analyzed in the arriveBy search that @optionsome noted.

I'll post some screenshots of the remaining weight estimates of vertices explored during the AStar search in various code states. Each vertex that was explored is shown as a dot and is color-coded with yellow meaning a remaining weight of 0 and then colors white to blue where white is lower remaining weight and blue is higher remaining weight.

Shown below is the vertices explored during the AStar search in the dev-1.x branch:

<img width="860" alt="Screen Shot 2019-03-11 at 5 02 29 PM" src="https://user-images.githubusercontent.com/3112493/54165807-82c2f300-441f-11e9-8f4d-0d6e62d29637.png">

With the current InterleavedBidirectionalHeuristic implementation in dev-1.x, 16,461 vertices were explored.

Including all of the changes to this heuristic up to commit https://github.com/opentripplanner/OpenTripPlanner/pull/2677/commits/61a28c0cad658203213761972846a9099e54fd0b, the vertices explored then looked like this:

<img width="854" alt="Screen Shot 2019-03-11 at 4 04 16 PM" src="https://user-images.githubusercontent.com/3112493/54163780-5c995500-4417-11e9-9b79-620136a55e8c.png">

With https://github.com/opentripplanner/OpenTripPlanner/pull/2677/commits/61a28c0cad658203213761972846a9099e54fd0b, 113,442 vertices were explored.

In the debugging I was doing, I was finding that with larger maxWalkDistance values for requests with the `BICYCLE` mode and the previous implementation of returning 0 for the estimated remaining weight of the pre-transit vertcies, the weights of states at vertices that hadn't boarded transit were not increasing fast enough for the transit vertices to rise to the top of the priority queue. This InterleavedBidirectionalHeuristic does work very well and as intended for transit+walk requests, but with so many vertices being explored in certain requests with the `BICYCLE` mode and especially the `CAR` mode, I decided to try adding a Euclidean heuristic for the pre-transit vertices in special cases where it is likely that many thousands of extra vertices could be explored.

I have gone ahead and implemented and refined this in the various commits after https://github.com/opentripplanner/OpenTripPlanner/pull/2677/commits/61a28c0cad658203213761972846a9099e54fd0b. This is what the remaining weight estimates for the vertices look like now:

<img width="841" alt="Screen Shot 2019-03-11 at 4 07 26 PM" src="https://user-images.githubusercontent.com/3112493/54163903-d3365280-4417-11e9-9eee-6858e5414985.png">

In this change (as of https://github.com/opentripplanner/OpenTripPlanner/pull/2677/commits/e62e6d4cb4a9b8dd49923863cd734c573d910d3d), 38,736 vertices were explored.

As can be seen, the number of vertices explored increases from dev-1.x but is less than what was observed without the euclidean addition to this heuristic. I ran this through the tests of the HSL fork of OTPQA and was able to get all tests passing for `WALK,TRANSIT`, `BICYCLE,TRANSIT` and `BICYCLE_RENT,TRANSIT` queries. Unfortunately, it appears that `CAR_PARK,TRANSIT` is broken in dev-1.x with the provided OSM data for Helsinki, so I wasn't able to test that mode combination. For the modes with bicycling these improvements lead to only a slight degradation in request time, but yield more itineraries.

A cautionary note is that this introduces a bias for vertices that are geographically closer to the target. It is possible for this implementation to lead the path search mostly towards the target when in fact a faster alternative exists by first going a certain distance in the opposite direction of the target.

After thinking more about how this heuristic is works, I believe that I have yet another idea for making this heuristic even better. As noted in a previous comment in this PR, I have been noticing that this heuristic explores a lot of the transit network. I'm thinking that it would be possible to allow the doSomeWork method to explore the pre-transit vertices as well and update remaining weight estimates which I am hypothesizing would result in more realistic remaining weight estimates, less vertices that would need to be explored, no need for the euclidean heuristic and faster response times.

optionsome commented:
@evansiroky I read through your comment and checked the latest changes, and it seems like the changes were in the right direction.

Have you decided yet if you are going to do more changes or should we review this more thoroughly now?

evansiroky commented:
I went ahead and tried implementing the changes I mentioned, but it just wasn't able to produce as good as results as the Euclidean heuristic for the pre-transit vertices. I did save my work into the [ibh-backward-experiment](https://github.com/opentripplanner/OpenTripPlanner/tree/ibh-backward-experiment) branch.

So I guess this current implementation is the best I can do for now.

optionsome commented on src/main/java/org/opentripplanner/routing/algorithm/strategies/InterleavedBidirectionalHeuristic.java:
```diff
@@ -87,6 +170,20 @@
+
+    // A threshold in meters beyond which the pre-transit euclidean heuristic is applied in order
+    // to save time with trip calculations
+    private final double PRE_TRANSIT_EUCLIDEAN_WALK_DISTANCE_THRESHOLD = 5000; // TODO determine a good value empirically
```
This value seems to be ok but perhaps a more optimized value could be found through testing but I didn't have time to do that now and maybe someone else has a better test setup for that purpose. It is also fine for me if this pull request gets merged with the current value.

evansiroky commented:
Thanks, @optionsome for the approval! I'm very happy to have earned it and I believe this PR is much better given our discussion and additional changes.

evansiroky commented:
@drewda Ready to merge?

evansiroky commented:
@optionsome and @fpurcell I just updated the changelog and now your reviews have been dismissed! Can you please reapprove this PR?
